openfisca_core/simulation_builder.py

```diff
@@ -159,9 +165,37 @@ def build_default_simulation(self, tax_benefit_system, count = 1, **kwargs):
             entity.ids = np.array(range(count))
             if not entity.is_person:
                 entity.members_entity_id = entity.ids  # Each person is its own group entity
-                entity.members_role = entity.filled_array(entity.flattened_roles[0])
         return simulation
 
+    def create_entities(self, tax_benefit_system):
+        self.entities_instances = tax_benefit_system.instantiate_entities()
+
+    def declare_person_entity(self, person_singular, persons_ids: Iterable):
+        person_instance = self.entities_instances[person_singular]
+        person_instance.ids = np.array(list(persons_ids))
+        person_instance.count = len(person_instance.ids)
+
+        self.persons_plural = person_instance.plural
+
+    def declare_entity(self, entity_singular, entity_ids: Iterable):
+        entity_instance = self.entities_instances[entity_singular]
+        entity_instance.ids = np.array(list(entity_ids))
+        entity_instance.count = len(entity_instance.ids)
+        return entity_instance
+
+    def nb_persons(self, entity_singular, role = None):
+        return self.entities_instances[entity_singular].nb_persons(role = role)
+
+    def join_with_persons(self, group_instance, persons_group_assignment, roles: Iterable[str]):
+        group_sorted_indices = np.unique(persons_group_assignment, return_inverse = True)[1]
+        group_instance.members_entity_id = np.argsort(group_instance.ids)[group_sorted_indices]
+
+        if roles is not None:
```

Comment:
If `roles` is now a mandatory argument of the method, is this condition still relevant?

Reply:
The *argument* being mandatory is not the same as the *data* being mandatory (on this commit you can still pass None). I've pushed a commit to move in this latter direction but I'm not 100% convinced it's a good idea, all of a sudden to input CSV into OpenFisca you have to know substantially more about the model and encode it into your data "cleanly".